File: articles/critical-periods-sensitive-periods-and-the-timing-of-development/go/timing_summary.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"strconv"
)

func findColumn(header []string, name string) int {
	for i, col := range header {
		if col == name {
			return i
		}
	}
	return -1
}

func main() {
	if len(os.Args) < 2 {
		log.Fatal("Usage: go run timing_summary.go data/developmental_timing_panel.csv")
	}

	file, err := os.Open(os.Args[1])
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatal(err)
	}

	if len(rows) < 2 {
		log.Fatal("CSV has no data rows")
	}

	header := rows[0]
	criticalIdx := findColumn(header, "critical_outcome")
	sensitiveIdx := findColumn(header, "sensitive_outcome")
	multiIdx := findColumn(header, "multi_window_outcome")

	if criticalIdx < 0 || sensitiveIdx < 0 || multiIdx < 0 {
		log.Fatal("Required outcome columns not found")
	}

	count := 0
	criticalSum := 0.0
	sensitiveSum := 0.0
	multiSum := 0.0

	for _, row := range rows[1:] {
		critical, err1 := strconv.ParseFloat(row[criticalIdx], 64)
		sensitive, err2 := strconv.ParseFloat(row[sensitiveIdx], 64)
		multi, err3 := strconv.ParseFloat(row[multiIdx], 64)

		if err1 != nil || err2 != nil || err3 != nil {
			continue
		}

		count++
		criticalSum += critical
		sensitiveSum += sensitive
		multiSum += multi
	}

	if count == 0 {
		log.Fatal("No valid rows found")
	}

	fmt.Printf("Rows analyzed: %d\n", count)
	fmt.Printf("Mean critical outcome: %.4f\n", criticalSum/float64(count))
	fmt.Printf("Mean sensitive outcome: %.4f\n", sensitiveSum/float64(count))
	fmt.Printf("Mean multi-window outcome: %.4f\n", multiSum/float64(count))
}

File: articles/critical-periods-sensitive-periods-and-the-timing-of-development/cpp/timing_summary.cpp
#include <fstream>
#include <iostream>
#include <sstream>
#include <string>
#include <vector>

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> values;
    std::stringstream ss(line);
    std::string item;

    while (std::getline(ss, item, ',')) {
        values.push_back(item);
    }

    return values;
}

int find_column(const std::vector<std::string>& header, const std::string& name) {
    for (size_t i = 0; i < header.size(); ++i) {
        if (header[i] == name) {
            return static_cast<int>(i);
        }
    }
    return -1;
}

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cerr << "Usage: timing_summary data/developmental_timing_panel.csv\n";
        return 1;
    }

    std::ifstream file(argv[1]);

    if (!file.is_open()) {
        std::cerr << "Unable to open CSV file\n";
        return 1;
    }

    std::string line;
    std::getline(file, line);

    auto header = split_csv_line(line);
    int critical_col = find_column(header, "critical_outcome");
    int sensitive_col = find_column(header, "sensitive_outcome");
    int multi_col = find_column(header, "multi_window_outcome");

    if (critical_col < 0 || sensitive_col < 0 || multi_col < 0) {
        std::cerr << "Required outcome columns not found\n";
        return 1;
    }

    long count = 0;
    double critical_sum = 0.0;
    double sensitive_sum = 0.0;
    double multi_sum = 0.0;

    while (std::getline(file, line)) {
        auto values = split_csv_line(line);

        if (
            static_cast<int>(values.size()) <= critical_col ||
            static_cast<int>(values.size()) <= sensitive_col ||
            static_cast<int>(values.size()) <= multi_col
        ) {
            continue;
        }

        critical_sum += std::stod(values[critical_col]);
        sensitive_sum += std::stod(values[sensitive_col]);
        multi_sum += std::stod(values[multi_col]);
        count += 1;
    }

    if (count == 0) {
        std::cerr << "No valid rows found\n";
        return 1;
    }

    std::cout << "Rows analyzed: " << count << "\n";
    std::cout << "Mean critical_outcome: " << critical_sum / count << "\n";
    std::cout << "Mean sensitive_outcome: " << sensitive_sum / count << "\n";
    std::cout << "Mean multi_window_outcome: " << multi_sum / count << "\n";

    return 0;
}
